API: add antialiased to interpolation-stage in image

jklymak opened the pull request:
UPDATE: 6 July 2024:

This PR changes the default *interpolation_stage* in imshow (and friends) to 'auto'.

It now _also_ changes the default *interpolation* from 'antialiased' to 'auto'.  

'auto' follows *interpolation* 'auto' in behaving differently if the image is upsampled by more than a factor of three or if it is weakly upsampled or downsampled.  If upsampled, *interpolation_stage* will happen at the data stage, like the previous default.  If downsampled, *interpolation_stage* has been changed to happen at the RGBA stage.    

When upsampling and using a smoother on the pixels, interpolation in RGBA space leads to non-sensical colors, and was the original motivation behind https://github.com/matplotlib/matplotlib/pull/5718.  Hence upsampling continues to happen in data space as has been the default since 2.0

![sphx_glr_image_antialiasing_010_2_00x-2](https://github.com/matplotlib/matplotlib/assets/1562854/53c4127c-9897-4496-9cea-fdbfb4453451)

Conversely, downsampling should have an anti-aliasing filter applied, but this leads to all sorts of strange effects if applied at the data stage (eg, https://github.com/matplotlib/matplotlib/issues/21167, and many other fixes since #5718 went in).  The new behaviour is to sidestep all these issues and do the interpolation in RGBA space if downsampling.  

![sphx_glr_image_antialiasing_008_2_00x](https://github.com/matplotlib/matplotlib/assets/1562854/84cc2673-333d-45d7-afdd-9139546c005b)

See the much longer description at https://output.circle-artifacts.com/output/job/edd10dae-2a13-48cf-af33-dca97ed402eb/artifacts/0/doc/build/html/gallery/images_contours_and_fields/image_antialiasing.html#sphx-glr-gallery-images-contours-and-fields-image-antialiasing-py

anntzer commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -45,71 +40,213 @@
+#
+# The default for the *interpolation* keyword argument is 'antialiased' which
+# will choose a Hanning filter if the image is being down-sampled or up-sampled
+# by less than a factor of three.  The default *interploation_stage* keyword
```
typo: "interploation"

anntzer commented on lib/matplotlib/image.py:
```diff
@@ -421,7 +421,21 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
+                if (dispx < 3) or (dispy < 3):
+                    interpolation_stage = 'rgba'
+                else:
+                    interpolation_stage = 'data'
```
I'm not sure there's a really sane default in the (pathological) case of an image being stretched along x and squeezed along y, but I'd have thought that the "neutral" solution would be to set the cutoff at `dispx * dispy < 1` (or `< 3`)?  i.e. rgba if there's fewer total pixels at the end than at the beginning, data, if there's more?

jklymak replied:
I'm erring on the side of using 'rgba' if any down-sampling is occurring - 'rgba' works exactly the same as 'data' if there is up-sampling and *interpolation* is 'nearest', which is the default.  The only issue with 'rgba' happens when the user wants to smooth upsampled pixels, and specifies 'sinc' or 'bilinear' (etc) smoothing.   Given that, I'm willing to ask such users to definitively specify 'data' if they are on the cusp of downsampling and getting weird artifacts.

tacaswell replied:
can we stuff this and the kernel selection into a helper function so that they do not get out-of-sync?

jklymak replied:
Yeah, I tried that but the kernel selection is buried at a different level in the code (in `_resample`) with different ways of getting the data and figure size, so I took the easy way out and quasi repeated the logic here.  I think a reasonable follow up is simplifying `_resample` and this part of the code base.

anntzer commented:
As briefly mentioned during the call, perhaps the option could be named in a way that does not preclude the future addition of an interpolation_stage="norm" ("normed", i.e. post-normalized data) option, and thus also of interpolation_stage="norm" when upsampling / "rgba" when downsampling.  For example the mixed options could be named `"up:data,down:rgba"` (so that `"up:norm,down:rgba"` can also be supported in the future)?  (Exact names up to bikeshedding...)

jklymak commented:
I don't think that is a problem. However we still need a default, and I think using 'antialiased', the same as the *interpolation* keyword, default makes sense.  Would it be a problem for that to by synonymous with 'up:data;down:rgba' if we wanted to offer that?

anntzer commented:
I guess the question is whether we'd be happy to change the meaning of "antialiased" to mean "up:norm,down:rgba" (which I consider a saner default) once "norm" comes in?

jklymak commented:
Oh, I see, sorry didn't catch on to that.  

Thats always an interesting question if one interpolates linearly or in log space.  In my field, people sometimes interpolate in log space, and then say incorrect things about the (arithmetic) mean.  Though I agree that there are aesthetic reasons to interpolate in log space depending on the problem.  

If we were to add a norm stage of interpolation, as a default, I would tend to err on the side of simple averaging (eg `up:data`) rather than after normalization.  At the lowest level it's easiest to explain.  However, if there is a consensus that we should default to interpolating after the norm, maybe we should try and put that in during the same cycle as this.

anntzer commented:
> maybe we should try and put that in during the same cycle as this.

OK; I guess this is not going to make it to 3.9(?)  (If it does, let's just go with "antialiasing" for now and explicitly mark this name as experimental (...but default) and reserve the right to change it later.)
Assuming this goes in 3.10 we should also get the stop-remapping-to-0/1 in as well, and probably first, as that should make interpolation_stage="norm" somewhat easier to implement (_make_image is already super unwieldy).

tacaswell commented on doc/api/next_api_changes/behavior/28061-JMK.rst:
```diff
@@ -0,0 +1,14 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The *interpolation_stage* keyword argument `~.Axes.imshow` has a new default
+value 'antialiased'.  For images that are down-sampled or up-sampled less than
+a factor of three, image interpolation will occur in 'rgba' space.  For images
```
```suggestion
value 'antialiased'.  For images that are up-sampled less than a factor of 
three or down-sampled , image interpolation will occur in 'rgba' space.  For images
```

I read it as "(down- or up-sampled) by less than a factor of 3" which is not what is meant.

tacaswell commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -1,34 +1,29 @@
+<https://en.wikipedia.org/wiki/Image_scaling>`_ the data or image to fit.  If
+the data array is larger than the number of pixels allotted in the image file,
+then the image will be "down-sampled" and image information will be lost.
+Conversely, if the data array is smaller than the number of pixels then each
```
```suggestion
Conversely, if the data array is smaller than the number of output pixels then each
```

tacaswell commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -45,71 +40,213 @@
+# white regions to the image that are not in the original data, both in the
+# alternating bands on the left hand side of the image, and in the boundary
+# between the red and blue of the large circles in the middle of the image.
+# The interpolation at the 'rgba' stage is more natural, with the alternating
```
```suggestion
# The interpolation at the 'rgba' stage has a different artifact, with the alternating
```

tacaswell commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -45,71 +40,213 @@
     ax.set_title(f"interpolation='{interp}'")
+
+# %%
+# A final example shows the desirability of performing the anti-aliasing at
+# the RGBA stage.  In the following, the data in the upper 100 rows is exactly
```
```suggestion
# A final example shows the desirability of performing the anti-aliasing at
# the RGBA stage when using non-trivial interpolation kernels.  In the following,
# the data in the upper 100 rows is exactly
```

tacaswell commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -45,71 +40,213 @@
+# or over- colors. This can be avoided if you do not use an anti-aliasing filter
+# (*interpolation* set set to 'nearest'), however, that makes the part of the
+# data susceptible to Moiré patterns much worse (second panel).  Therefore, we
+# recommend the default *interpolation* of 'hanning'/'antialiased', and
+# *interpolation_stage* of 'rgba'/'antialiased' for most down-sampling
+# situations (last panel).
```
```suggestion
# recommend the either using default *interpolation* of 'hanning'/'antialiased', 
# and *interpolation_stage* of 'rgba'/'antialiased' for most down-sampling
# situations (last panel) or *interpolation* of 'nearest' and *interpolation_stage*
# of 'data'.
```

jklymak replied:
This seems more confusing than helpful.  We've already said what the different arguments do, and we have a default that we recommend.

tacaswell commented on galleries/examples/images_contours_and_fields/image_antialiasing.py:
```diff
@@ -45,71 +40,213 @@
+#
+# The second method is to exactly match the size of your axes to the size of
+# your data. in the following, the figure is exactly 2 inches by 2 inches, and
+# the dpi is 200, so the 400x400 data is not resampled at all. If you download
```
I think there is some normal/high dpi magic going on with saved images so it is either 100 or 200 based on the screen the reader has.

jklymak replied:
OK< I put the dpi in as hard coded.  Not sure what happens for a low dpi screen if I do this.

tacaswell commented on lib/matplotlib/image.py:
```diff
@@ -552,7 +566,7 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                     output = self.norm(resampled_masked)
             else:
-                if A.ndim == 2:  # _interpolation_stage == 'rgba'
+                if A.ndim == 2:  # interpolation_stage == 'rgba'
```
Can we do this as an `elif` and add an `else` that raises?  Or do we do sufficient validation else where that we are sure it will always be one of the valid keys when we get here?

jklymak replied:
Yeah, keys are validated elsewhere, and this needs to be an else because it covers both ndim>2 and rgba.

tacaswell commented on lib/matplotlib/tests/test_image.py:
```diff
@@ -1462,6 +1459,45 @@ def test_rgba_antialias():
+
+    grid = np.random.rand(4000, 4000)
+    ax = fig_ref.subplots()
+    ax.imshow(grid, interpolation='antialiased', cmap='viridis',
```
```suggestion
    ax.imshow(grid, interpolation='hanning', cmap='viridis',
```

leave nothing to chance?

jklymak replied:
Well the opposite - I'd like this to fail if we change the default, so at least we do it purposefully?

tacaswell replied:
fair.

tacaswell commented:
Overall 👍🏻 thank you for putting in the leg work on this @jklymak !

I think we should target this for 3.10 to give us enough time to get early users and to get the additional changes proposed by @anntzer in.

I left word choice edits that I think soften the prose a bit without compromising the core meaning.

anntzer commented:
(I didn't mention this but the docs are awesome.)

tacaswell commented:
> (I didn't mention this but the docs are awesome.)

I completely agree!  Should have lead with that instead of a bunch of picky edits.

jklymak commented:
Edits all seem good, I'll put those in.  I think 3.10 is completely reasonable, and happy to work on the rest of the chain, but maybe in a few weeks.  The todos are to simplify the 0-1 clipping issues, and then add a Norm interpolation stage (which would be easier if the 0-1 clipping issue were cleared up.  

I think this could go in when I make the edits above, and then update if we get the rest in.  Some of the docs here will need to change and expand anyhow, and it would be easier to start with what is here than amalgamate

timhoffm commented on doc/api/next_api_changes/behavior/28061-JMK.rst:
```diff
@@ -0,0 +1,14 @@
+the new default.  However, the new default also avoids floating point artifacts
+at sharp boundaries in a colormap when down-sampling.
+
+The previous behavior can achieved by changing :rc:`image.interpolation_stage`.
```
```suggestion
The previous behavior can achieved by setting :rc:`image.interpolation_stage` or
the *interpolation_stage* parameter in `~.axes.Axes.imshow` to 'data'.
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5787,11 +5787,14 @@ def imshow(self, X, cmap=None, norm=None, *, aspect=None,
-            is carried out on the data provided by the user.  If 'rgba', the
-            interpolation is carried out after the colormapping has been
-            applied (visual interpolation).
+        interpolation_stage : {'antialiased', 'data', 'rgba'}, default: 'antialiased'
```
```suggestion
        interpolation_stage : {'antialiased', 'data', 'rgba'}, default: :rc:`interpolation_stage`
```

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5787,11 +5787,14 @@ def imshow(self, X, cmap=None, norm=None, *, aspect=None,
+        interpolation_stage : {'antialiased', 'data', 'rgba'}, default: 'antialiased'
+            If 'data', interpolation is carried out on the data provided by the user.
+            If 'rgba', the interpolation is carried out in RGBA-space after the
+            color-mapping has been applied (visual interpolation).  If 'antialiased',
```
Is 'antialiased' the correct name?

- The other two options 'data' and 'rgba' refer to types of data in the processing pipeline. In contrast 'antialiased' is an effect.
- From https://en.wikipedia.org/wiki/Spatial_anti-aliasing:

  > In digital signal processing, spatial anti-aliasing is a technique for minimizing the distortion artifacts (aliasing) when representing a high-resolution image at a lower resolution.

  In my understanding anti-aliasing is only connected to downsampling. It's thus surprising that a mode 'antialiased' tries to
  cover down-sampling and up-sampling.

Suggestion: Simply use 'auto', which hints at "use the best (from existing modes 'data', 'rgba') for the given context".

jklymak replied:
'antialiased' is meant to be the same for both *interpolation* and *interpolation_stage* keyword arguments.  The analogy is with most viewers, which have an "antialiased" toggle, regardless of whether the image is actually up or downsampled.

timhoffm replied:
I see. But still the name does not speak to me. I recommend to get some opinions from core devs. If they are fine with the name, I'll give in.

greglucas replied:
I think "auto" would be better for both as it implies some magic selection is going on for the user to determine what is best. I agree, it seems like we coudl point "antialiased" -> "auto" if a user inputs "antialiased" and just rewrite the docs to make everything show "auto" now.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5787,11 +5787,14 @@ def imshow(self, X, cmap=None, norm=None, *, aspect=None,
-            interpolation is carried out after the colormapping has been
-            applied (visual interpolation).
+        interpolation_stage : {'antialiased', 'data', 'rgba'}, default: 'antialiased'
+            If 'data', interpolation is carried out on the data provided by the user.
+            If 'rgba', the interpolation is carried out in RGBA-space after the
```
Optionally, give a hint when these are suited:

'data' is well suited for creating intermediate values when up-scaling. 'rgba' reduces anti-aliasing when down-scaling.

jklymak replied:
Does the next sentence explaining the default basically gives the hint?

timhoffm replied:
Yes, it gives a hint, but only a hint. There could be more and clearer information.

It says which is used in which case. One may assume that these are the best respective options for the cases, but it's not explicitly stated (There could be other conditions involved in the decision for 'antialiased'). Also not stated is the respective benefit "suited for creating intermediate values" / "reduce anti-aliasing".

If we only had "antialiased", that would be ok, because it essentially communicates "you get some nice rendering". But if you have the other options, you may want to know what they do exactly, when you want to choose them and what the advantage/disadadvange is.

jklymak replied:
OK, Changed to 
```
            If 'data', interpolation is carried out on the data provided by the user,
            useful if interpolating between pixels during upsampling.
            If 'rgba', the interpolation is carried out in RGBA-space after the
            color-mapping has been applied, useful if downsampling and combining
            pixels visually. The default 'antialiased' is appropriate for most
            applications where 'rgba' is used when downsampling, or upsampling at a
            rate less than 3, and 'data' is used when upsampling at a higher rate.
            See :doc:`/gallery/images_contours_and_fields/image_antialiasing` for
            a discussion of image antialiasing.

```

timhoffm reviewed (commented):
Awesome improvement! 🚀

greglucas commented on lib/matplotlib/tests/test_image.py:
```diff
@@ -1606,3 +1642,84 @@ def test_non_transdata_image_does_not_touch_aspect():
+    axs[0].set_title("interpolation='antialiased', stage='rgba'")
+    axs[0].imshow(np.triu(img), cmap=cm, norm=norm, interpolation_stage='rgba')
+
+    # Should be same as previous
```
Can you do a figure reference/test comparison test here instead with these two axes?

jklymak commented:
OK, changed the default from 'antialiased' to 'auto'

Also changed *interpolation* default to 'auto' from 'antialiased'.  Kept 'antialiased' for back compatibility, but should behave exactly as 'auto'.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -5832,11 +5832,16 @@ def imshow(self, X, cmap=None, norm=None, *, aspect=None,
-            interpolation is carried out after the colormapping has been
-            applied (visual interpolation).
+        interpolation_stage : {'auto', 'data', 'rgba'}, default: 'auto'
+            If 'data', interpolation is carried out on the data provided by the user,
+            useful if interpolating between pixels during upsampling.
+            If 'rgba', the interpolation is carried out in RGBA-space after the
+            color-mapping has been applied, useful if downsampling and combining
+            pixels visually. The default 'auto' is appropriate for most
+            applications where 'rgba' is used when downsampling, or upsampling at a
+            rate less than 3, and 'data' is used when upsampling at a higher rate.
+            See :doc:`/gallery/images_contours_and_fields/image_antialiasing` for
+            a discussion of image antialiasing.
```
Optional: I like bullet-listing
```suggestion
            Supported values:
            
            - 'data': Interpolation is carried out on the data provided by the user
              This is useful if interpolating between pixels during upsampling.
            - 'rgba': The interpolation is carried out in RGBA-space after the
              color-mapping has been applied. This is useful if downsampling and
              combining pixels visually.
            - 'auto': Select a suitable interpolation stage automatically. This uses
              'rgba' when downsampling, or upsampling at a rate less than 3, and
              'data' when upsampling at a higher rate.
              
            See :doc:`/gallery/images_contours_and_fields/image_antialiasing` for
            a discussion of image antialiasing.
```

timhoffm commented on doc/api/next_api_changes/behavior/28061-JMK.rst:
```diff
@@ -0,0 +1,22 @@
+The *interpolation_stage* keyword argument `~.Axes.imshow` has a new default
+value 'auto'.  For images that are up-sampled less than a factor of
+three or down-sampled, image interpolation will occur in 'rgba' space.  For images
+that are up-sampled by more than a factor of 3, then image interpolation occurs
```
```suggestion
that are up-sampled by a factor of 3 or more, then image interpolation occurs
```

timhoffm commented on lib/matplotlib/tests/test_png.py:
```diff
@@ -20,7 +20,9 @@ def test_pngsuite():
             # keep grayscale images gray
             cmap = cm.gray
-        plt.imshow(data, extent=(i, i + 1, 0, 1), cmap=cmap)
+        # use the old default data interpolation stage
```
```suggestion
        # Using the old default data interpolation stage lets us
        # continue to use the existing reference image
```

timhoffm reviewed (approved):
Only minor documentation suggestions.

greglucas commented:
Thanks for putting in the work to make these changes, @jklymak!

jklymak commented:
Thanks for the help - the is a relatively big change, but hopefully will not cause too much blowback and reduce bug reports for weird under sampled images...

tacaswell commented:
Thank you everyone!
